Implement error handling and recovery

## Summary

Implement comprehensive error handling and recovery system for Yellowwood with graceful degradation, error boundaries for git operations, file system permission handling, and user-friendly notifications. Ensures the application remains stable and provides helpful feedback when things go wrong.

## Context

Yellowwood interacts with several external systems that can fail: git operations, file system access, configuration loading, and process execution (editor opening). Without proper error handling, these failures would crash the application or leave it in an inconsistent state. This issue implements a defensive error handling strategy that:

1. **Catches errors at boundaries** - Git, filesystem, and config operations are wrapped in try/catch
2. **Degrades gracefully** - Missing git, permission errors, or invalid configs disable features rather than crash
3. **Notifies users clearly** - Every error produces a helpful notification explaining what happened
4. **Recovers automatically** - Transient errors (network, file locks) are retried with exponential backoff
5. **Logs appropriately** - Errors are logged for debugging without cluttering the UI

**Current State:**
- Basic error handling exists in some utilities (config.ts, git.ts)
- No centralized error handling strategy
- No error recovery mechanisms
- Limited user feedback on errors
- No error boundaries around critical operations

**What This Issue Delivers:**
- Centralized error handling utilities
- Error boundaries for git, filesystem, and config operations
- Retry logic with exponential backoff for transient errors
- Comprehensive user notifications with actionable messages
- Graceful degradation for missing dependencies
- Error logging with appropriate severity levels

**How This Will Be Used:**

```typescript
// In useGitStatus hook
const statusMap = await withGitErrorBoundary(
  async () => await getGitStatus(cwd),
  { fallback: new Map(), notify: true }
);

// In file tree builder
const files = await withFileSystemErrorBoundary(
  async () => await buildFileTree(rootPath),
  { fallback: [], notify: true }
);

// In file watcher
await withRetry(
  async () => await setupWatcher(path),
  { maxRetries: 3, backoff: 'exponential', notify: false }
);
```

## Deliverables

### Code Changes

**Files to Create:**
- `src/utils/errorHandling.ts` - Error handling utilities and error boundaries
- `src/utils/errorTypes.ts` - Custom error types with context
- `src/utils/logger.ts` - Structured logging utilities

**Files to Modify:**
- `src/utils/git.ts` - Add error boundaries and notifications
- `src/utils/config.ts` - Improve error messages and recovery
- `src/utils/fileWatcher.ts` - Add retry logic for watcher setup
- `src/hooks/useGitStatus.ts` - Handle git errors gracefully
- `src/hooks/useFileTree.ts` - Handle filesystem errors
- `src/App.tsx` - Wrap critical operations in error boundaries

### Implementation Details

**Step 1: Define Custom Error Types**

Create `src/utils/errorTypes.ts`:

```typescript
/**
 * Base error class for all Yellowwood errors.
 * Includes context for better debugging and user messages.
 */
export class YellowwoodError extends Error {
  constructor(
    message: string,
    public readonly context?: Record<string, unknown>,
    public readonly cause?: Error
  ) {
    super(message);
    this.name = this.constructor.name;
    Error.captureStackTrace?.(this, this.constructor);
  }
}

/**
 * Git operation failed (repository not found, git not installed, command failed)
 */
export class GitError extends YellowwoodError {
  constructor(message: string, context?: Record<string, unknown>, cause?: Error) {
    super(message, context, cause);
  }
}

/**
 * File system operation failed (permission denied, file not found, read error)
 */
export class FileSystemError extends YellowwoodError {
  constructor(message: string, context?: Record<string, unknown>, cause?: Error) {
    super(message, context, cause);
  }
}

/**
 * Configuration loading or validation failed
 */
export class ConfigError extends YellowwoodError {
  constructor(message: string, context?: Record<string, unknown>, cause?: Error) {
    super(message, context, cause);
  }
}

/**
 * External process execution failed (editor not found, command error)
 */
export class ProcessError extends YellowwoodError {
  constructor(message: string, context?: Record<string, unknown>, cause?: Error) {
    super(message, context, cause);
  }
}

/**
 * File watcher setup or operation failed
 */
export class WatcherError extends YellowwoodError {
  constructor(message: string, context?: Record<string, unknown>, cause?: Error) {
    super(message, context, cause);
  }
}

/**
 * Check if error is a specific Yellowwood error type
 */
export function isYellowwoodError(error: unknown): error is YellowwoodError {
  return error instanceof YellowwoodError;
}

/**
 * Check if error is a permission/access error (EACCES, EPERM)
 */
export function isPermissionError(error: unknown): boolean {
  if (!error || typeof error !== 'object') return false;
  const code = (error as NodeJS.ErrnoException).code;
  return code === 'EACCES' || code === 'EPERM';
}

/**
 * Check if error is a "not found" error (ENOENT)
 */
export function isNotFoundError(error: unknown): boolean {
  if (!error || typeof error !== 'object') return false;
  return (error as NodeJS.ErrnoException).code === 'ENOENT';
}

/**
 * Check if error is a transient error that might succeed on retry
 */
export function isTransientError(error: unknown): boolean {
  if (!error || typeof error !== 'object') return false;
  const code = (error as NodeJS.ErrnoException).code;
  // File busy, resource temporarily unavailable, etc.
  return ['EBUSY', 'EAGAIN', 'ETIMEDOUT', 'ECONNRESET'].includes(code || '');
}

/**
 * Extract user-friendly message from any error
 */
export function getUserMessage(error: unknown): string {
  if (isYellowwoodError(error)) {
    return error.message;
  }

  if (error instanceof Error) {
    return error.message;
  }

  return String(error);
}

/**
 * Extract technical details for logging
 */
export function getErrorDetails(error: unknown): Record<string, unknown> {
  const details: Record<string, unknown> = {
    message: getUserMessage(error),
  };

  if (error instanceof Error) {
    details.name = error.name;
    details.stack = error.stack;
  }

  if (isYellowwoodError(error)) {
    details.context = error.context;
    if (error.cause) {
      details.cause = getErrorDetails(error.cause);
    }
  }

  if (error && typeof error === 'object') {
    const nodeError = error as NodeJS.ErrnoException;
    if (nodeError.code) details.code = nodeError.code;
    if (nodeError.errno) details.errno = nodeError.errno;
    if (nodeError.syscall) details.syscall = nodeError.syscall;
    if (nodeError.path) details.path = nodeError.path;
  }

  return details;
}
```

**Step 2: Implement Structured Logger**

Create `src/utils/logger.ts`:

```typescript
/**
 * Structured logging utilities for Yellowwood.
 * Uses console.log/warn/error with consistent formatting.
 */

export type LogLevel = 'debug' | 'info' | 'warn' | 'error';

interface LogContext {
  [key: string]: unknown;
}

/**
 * Log a debug message (development only, filtered in production)
 */
export function logDebug(message: string, context?: LogContext): void {
  if (process.env.NODE_ENV === 'development') {
    console.log(`[DEBUG] ${message}`, context ? JSON.stringify(context, null, 2) : '');
  }
}

/**
 * Log an info message
 */
export function logInfo(message: string, context?: LogContext): void {
  console.log(`[INFO] ${message}`, context ? JSON.stringify(context, null, 2) : '');
}

/**
 * Log a warning message
 */
export function logWarn(message: string, context?: LogContext): void {
  console.warn(`[WARN] ${message}`, context ? JSON.stringify(context, null, 2) : '');
}

/**
 * Log an error message
 */
export function logError(message: string, error?: unknown, context?: LogContext): void {
  const errorDetails = error ? getErrorDetails(error) : undefined;
  console.error(
    `[ERROR] ${message}`,
    errorDetails ? JSON.stringify(errorDetails, null, 2) : '',
    context ? JSON.stringify(context, null, 2) : ''
  );
}

/**
 * Extract error details for logging
 */
function getErrorDetails(error: unknown): Record<string, unknown> {
  if (error instanceof Error) {
    return {
      name: error.name,
      message: error.message,
      stack: error.stack,
    };
  }

  return { value: String(error) };
}
```

**Step 3: Implement Error Boundaries and Retry Logic**

Create `src/utils/errorHandling.ts`:

```typescript
import type { Notification, NotificationType } from '../types/index.js';
import { logError, logWarn, logDebug } from './logger.js';
import {
  GitError,
  FileSystemError,
  ConfigError,
  ProcessError,
  WatcherError,
  isTransientError,
  isPermissionError,
  isNotFoundError,
  getUserMessage,
  getErrorDetails,
} from './errorTypes.js';

/**
 * Configuration for error boundaries
 */
interface ErrorBoundaryOptions<T> {
  fallback: T;                          // Value to return on error
  notify?: boolean;                     // Show notification to user
  notificationType?: NotificationType;  // Override notification type
  onError?: (error: unknown) => void;   // Custom error handler
  context?: Record<string, unknown>;    // Additional context for logging
}

/**
 * Configuration for retry logic
 */
interface RetryOptions {
  maxRetries?: number;           // Maximum number of retries (default: 3)
  backoff?: 'none' | 'linear' | 'exponential';  // Backoff strategy (default: exponential)
  baseDelay?: number;            // Base delay in ms (default: 100)
  maxDelay?: number;             // Maximum delay in ms (default: 5000)
  shouldRetry?: (error: unknown, attempt: number) => boolean;  // Custom retry predicate
  notify?: boolean;              // Show notification on final failure
  onRetry?: (error: unknown, attempt: number, delay: number) => void;  // Retry callback
}

/**
 * Wrap a git operation in an error boundary.
 * Catches git errors and provides fallback value.
 */
export async function withGitErrorBoundary<T>(
  operation: () => Promise<T>,
  options: ErrorBoundaryOptions<T>
): Promise<T> {
  try {
    return await operation();
  } catch (error) {
    const gitError = error instanceof GitError ? error : new GitError(
      'Git operation failed',
      options.context,
      error instanceof Error ? error : undefined
    );

    logError('Git error', gitError, options.context);

    if (options.notify) {
      // Notification handled by caller via onError callback
      options.onError?.(gitError);
    }

    return options.fallback;
  }
}

/**
 * Wrap a file system operation in an error boundary.
 * Handles permission errors, not found errors, and other filesystem failures.
 */
export async function withFileSystemErrorBoundary<T>(
  operation: () => Promise<T>,
  options: ErrorBoundaryOptions<T>
): Promise<T> {
  try {
    return await operation();
  } catch (error) {
    let fsError: FileSystemError;
    let userMessage = 'File system operation failed';

    if (isPermissionError(error)) {
      const path = (error as NodeJS.ErrnoException).path || 'unknown';
      userMessage = `Permission denied: Cannot access ${path}`;
    } else if (isNotFoundError(error)) {
      const path = (error as NodeJS.ErrnoException).path || 'unknown';
      userMessage = `File not found: ${path}`;
    }

    fsError = error instanceof FileSystemError ? error : new FileSystemError(
      userMessage,
      options.context,
      error instanceof Error ? error : undefined
    );

    const severity = isPermissionError(error) || isNotFoundError(error) ? 'warn' : 'error';
    if (severity === 'warn') {
      logWarn(userMessage, options.context);
    } else {
      logError(userMessage, fsError, options.context);
    }

    if (options.notify) {
      options.onError?.(fsError);
    }

    return options.fallback;
  }
}

/**
 * Wrap a config operation in an error boundary.
 * Validates and provides helpful messages for config errors.
 */
export async function withConfigErrorBoundary<T>(
  operation: () => Promise<T>,
  options: ErrorBoundaryOptions<T>
): Promise<T> {
  try {
    return await operation();
  } catch (error) {
    const configError = error instanceof ConfigError ? error : new ConfigError(
      'Configuration error',
      options.context,
      error instanceof Error ? error : undefined
    );

    logError('Config error', configError, options.context);

    if (options.notify) {
      options.onError?.(configError);
    }

    return options.fallback;
  }
}

/**
 * Retry an operation with exponential backoff.
 * Useful for transient errors like file locks or network issues.
 */
export async function withRetry<T>(
  operation: () => Promise<T>,
  options: RetryOptions = {}
): Promise<T> {
  const {
    maxRetries = 3,
    backoff = 'exponential',
    baseDelay = 100,
    maxDelay = 5000,
    shouldRetry = isTransientError,
    notify = true,
    onRetry,
  } = options;

  let lastError: unknown;
  let attempt = 0;

  while (attempt <= maxRetries) {
    try {
      return await operation();
    } catch (error) {
      lastError = error;
      attempt++;

      // Check if we should retry
      if (attempt > maxRetries || !shouldRetry(error, attempt)) {
        break;
      }

      // Calculate delay
      let delay = baseDelay;
      if (backoff === 'linear') {
        delay = baseDelay * attempt;
      } else if (backoff === 'exponential') {
        delay = Math.min(baseDelay * Math.pow(2, attempt - 1), maxDelay);
      }

      logDebug(`Retrying operation (attempt ${attempt}/${maxRetries}) after ${delay}ms`, {
        error: getUserMessage(error),
      });

      onRetry?.(error, attempt, delay);

      // Wait before retrying
      await new Promise(resolve => setTimeout(resolve, delay));
    }
  }

  // All retries exhausted
  const message = `Operation failed after ${attempt} attempts`;
  logError(message, lastError);

  if (notify) {
    // Caller should handle notification via thrown error
  }

  throw lastError;
}

/**
 * Safely execute an operation with comprehensive error handling.
 * Combines error boundary + retry logic + logging.
 */
export async function safeExecute<T>(
  operation: () => Promise<T>,
  options: {
    fallback: T;
    retry?: RetryOptions;
    boundary?: Omit<ErrorBoundaryOptions<T>, 'fallback'>;
  }
): Promise<T> {
  const boundaryOptions: ErrorBoundaryOptions<T> = {
    fallback: options.fallback,
    ...options.boundary,
  };

  return withFileSystemErrorBoundary(async () => {
    if (options.retry) {
      return await withRetry(operation, options.retry);
    }
    return await operation();
  }, boundaryOptions);
}

/**
 * Create a user-friendly notification from an error.
 */
export function createErrorNotification(
  error: unknown,
  defaultMessage: string = 'An error occurred'
): Notification {
  let message = getUserMessage(error);
  let type: NotificationType = 'error';

  // Downgrade severity for certain error types
  if (isPermissionError(error) || isNotFoundError(error)) {
    type = 'warning';
  }

  // If message is too technical, use default
  if (!message || message.length < 10) {
    message = defaultMessage;
  }

  return { message, type };
}

/**
 * Handle uncaught errors globally.
 * Should be set up in App.tsx on mount.
 */
export function setupGlobalErrorHandler(
  onError: (error: unknown) => void
): () => void {
  const handler = (event: ErrorEvent) => {
    event.preventDefault();
    logError('Uncaught error', event.error);
    onError(event.error);
  };

  const rejectionHandler = (event: PromiseRejectionEvent) => {
    event.preventDefault();
    logError('Unhandled promise rejection', event.reason);
    onError(event.reason);
  };

  // Note: These are browser APIs - not available in Node
  // For CLI apps, handle process-level errors differently
  if (typeof window !== 'undefined') {
    window.addEventListener('error', handler);
    window.addEventListener('unhandledrejection', rejectionHandler);

    return () => {
      window.removeEventListener('error', handler);
      window.removeEventListener('unhandledrejection', rejectionHandler);
    };
  }

  // For Node.js CLI apps
  const processErrorHandler = (error: Error) => {
    logError('Uncaught exception', error);
    onError(error);
  };

  const processRejectionHandler = (reason: unknown) => {
    logError('Unhandled rejection', reason);
    onError(reason);
  };

  process.on('uncaughtException', processErrorHandler);
  process.on('unhandledRejection', processRejectionHandler);

  return () => {
    process.off('uncaughtException', processErrorHandler);
    process.off('unhandledRejection', processRejectionHandler);
  };
}
```

**Step 4: Update Existing Utilities**

Modify `src/utils/git.ts` to use error boundaries:

```typescript
import { GitError } from './errorTypes.js';
import { logWarn } from './logger.js';

export async function isGitRepository(path: string): Promise<boolean> {
  try {
    const git: SimpleGit = simpleGit(path);
    await git.status();
    return true;
  } catch (error) {
    // Not a git repo or git not installed - this is expected, not an error
    logWarn('Git repository check failed', { path, error: (error as Error).message });
    return false;
  }
}

export async function getGitStatus(cwd: string): Promise<Map<string, GitStatus>> {
  const statusMap = new Map<string, GitStatus>();

  try {
    const git: SimpleGit = simpleGit(cwd);
    const status: StatusResult = await git.status();

    // ... existing mapping logic ...

  } catch (error) {
    // Git operation failed - wrap in GitError for better context
    throw new GitError(
      'Failed to get git status',
      { cwd },
      error instanceof Error ? error : undefined
    );
  }

  return statusMap;
}
```

Modify `src/utils/config.ts` to use error types:

```typescript
import { ConfigError } from './errorTypes.js';
import { logWarn, logError } from './logger.js';

export async function loadConfig(cwd: string = process.cwd()): Promise<YellowwoodConfig> {
  try {
    const explorer = cosmiconfig('yellowwood', { /* ... */ });
    const projectResult = await explorer.search(cwd);
    const projectConfig = projectResult?.config || {};

    const globalConfig = await loadGlobalConfig();
    const merged = mergeConfigs(DEFAULT_CONFIG, globalConfig, projectConfig);

    return validateConfig(merged);
  } catch (error) {
    if (error instanceof ConfigError) {
      throw error;
    }

    throw new ConfigError(
      'Failed to load configuration',
      { cwd },
      error instanceof Error ? error : undefined
    );
  }
}

function validateConfig(config: unknown): YellowwoodConfig {
  // ... existing validation ...

  if (errors.length > 0) {
    throw new ConfigError(
      `Config validation failed:\n${errors.join('\n')}`,
      { config }
    );
  }

  return c as YellowwoodConfig;
}
```

**Step 5: Integrate with App Component**

Modify `src/App.tsx` to use error handling:

```typescript
import { setupGlobalErrorHandler, createErrorNotification } from './utils/errorHandling.js';
import { withConfigErrorBoundary, withGitErrorBoundary } from './utils/errorHandling.js';

export function App({ rootPath = process.cwd() }: AppProps): JSX.Element {
  const [notification, setNotification] = useState<Notification | null>(null);
  const [config, setConfig] = useState<YellowwoodConfig>(DEFAULT_CONFIG);
  const [gitEnabled, setGitEnabled] = useState(true);

  // Set up global error handler
  useEffect(() => {
    const cleanup = setupGlobalErrorHandler((error) => {
      const errorNotif = createErrorNotification(error, 'An unexpected error occurred');
      setNotification(errorNotif);
    });

    return cleanup;
  }, []);

  // Load configuration with error boundary
  useEffect(() => {
    const loadConfigSafely = async () => {
      const loadedConfig = await withConfigErrorBoundary(
        async () => await loadConfig(rootPath),
        {
          fallback: DEFAULT_CONFIG,
          notify: true,
          onError: (error) => {
            setNotification({
              message: 'Failed to load configuration, using defaults',
              type: 'warning',
            });
          },
        }
      );

      setConfig(loadedConfig);
    };

    loadConfigSafely();
  }, [rootPath]);

  // Check git availability with error boundary
  useEffect(() => {
    const checkGit = async () => {
      const isRepo = await withGitErrorBoundary(
        async () => await isGitRepository(rootPath),
        {
          fallback: false,
          notify: false,
          onError: (error) => {
            setGitEnabled(false);
            setNotification({
              message: 'Git not available, git features disabled',
              type: 'info',
            });
          },
        }
      );

      setGitEnabled(isRepo);
    };

    checkGit();
  }, [rootPath]);

  // ... rest of component
}
```

### Tests

Create `tests/utils/test_errorHandling.ts`:

```typescript
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import {
  withGitErrorBoundary,
  withFileSystemErrorBoundary,
  withRetry,
  createErrorNotification,
} from '../../src/utils/errorHandling.js';
import { GitError, FileSystemError, isTransientError } from '../../src/utils/errorTypes.js';

describe('errorHandling', () => {
  describe('withGitErrorBoundary', () => {
    it('returns result when operation succeeds', async () => {
      const operation = vi.fn().mockResolvedValue('success');
      const result = await withGitErrorBoundary(operation, { fallback: 'fallback' });

      expect(result).toBe('success');
      expect(operation).toHaveBeenCalledTimes(1);
    });

    it('returns fallback when operation throws GitError', async () => {
      const operation = vi.fn().mockRejectedValue(new GitError('Git failed'));
      const result = await withGitErrorBoundary(operation, { fallback: 'fallback' });

      expect(result).toBe('fallback');
    });

    it('calls onError callback when notify is true', async () => {
      const operation = vi.fn().mockRejectedValue(new Error('Git failed'));
      const onError = vi.fn();

      await withGitErrorBoundary(operation, {
        fallback: 'fallback',
        notify: true,
        onError,
      });

      expect(onError).toHaveBeenCalledTimes(1);
      expect(onError).toHaveBeenCalledWith(expect.any(GitError));
    });

    it('wraps non-GitError in GitError', async () => {
      const operation = vi.fn().mockRejectedValue(new Error('Generic error'));
      const onError = vi.fn();

      await withGitErrorBoundary(operation, {
        fallback: 'fallback',
        notify: true,
        onError,
      });

      expect(onError).toHaveBeenCalledWith(expect.any(GitError));
    });
  });

  describe('withFileSystemErrorBoundary', () => {
    it('returns result when operation succeeds', async () => {
      const operation = vi.fn().mockResolvedValue({ data: 'test' });
      const result = await withFileSystemErrorBoundary(operation, { fallback: null });

      expect(result).toEqual({ data: 'test' });
    });

    it('returns fallback on permission error', async () => {
      const permError = Object.assign(new Error('EACCES'), { code: 'EACCES', path: '/root' });
      const operation = vi.fn().mockRejectedValue(permError);

      const result = await withFileSystemErrorBoundary(operation, { fallback: null });

      expect(result).toBe(null);
    });

    it('returns fallback on not found error', async () => {
      const notFoundError = Object.assign(new Error('ENOENT'), { code: 'ENOENT', path: '/missing' });
      const operation = vi.fn().mockRejectedValue(notFoundError);

      const result = await withFileSystemErrorBoundary(operation, { fallback: [] });

      expect(result).toEqual([]);
    });

    it('provides helpful message for permission errors', async () => {
      const permError = Object.assign(new Error('EACCES'), { code: 'EACCES', path: '/root/file' });
      const operation = vi.fn().mockRejectedValue(permError);
      const onError = vi.fn();

      await withFileSystemErrorBoundary(operation, {
        fallback: null,
        notify: true,
        onError,
      });

      expect(onError).toHaveBeenCalledWith(expect.any(FileSystemError));
      const error = onError.mock.calls[0][0] as FileSystemError;
      expect(error.message).toContain('Permission denied');
      expect(error.message).toContain('/root/file');
    });
  });

  describe('withRetry', () => {
    beforeEach(() => {
      vi.useFakeTimers();
    });

    afterEach(() => {
      vi.restoreAllMocks();
    });

    it('succeeds on first attempt', async () => {
      const operation = vi.fn().mockResolvedValue('success');
      const result = await withRetry(operation);

      expect(result).toBe('success');
      expect(operation).toHaveBeenCalledTimes(1);
    });

    it('retries on transient error and eventually succeeds', async () => {
      const transientError = Object.assign(new Error('EBUSY'), { code: 'EBUSY' });
      const operation = vi.fn()
        .mockRejectedValueOnce(transientError)
        .mockRejectedValueOnce(transientError)
        .mockResolvedValueOnce('success');

      const promise = withRetry(operation, { maxRetries: 3, baseDelay: 100 });

      // Fast-forward through retries
      await vi.advanceTimersByTimeAsync(100); // First retry
      await vi.advanceTimersByTimeAsync(200); // Second retry (exponential backoff)

      const result = await promise;

      expect(result).toBe('success');
      expect(operation).toHaveBeenCalledTimes(3);
    });

    it('throws after max retries exhausted', async () => {
      const transientError = Object.assign(new Error('EBUSY'), { code: 'EBUSY' });
      const operation = vi.fn().mockRejectedValue(transientError);

      const promise = withRetry(operation, { maxRetries: 2, baseDelay: 100 });

      // Fast-forward through all retries
      await vi.advanceTimersByTimeAsync(100);
      await vi.advanceTimersByTimeAsync(200);

      await expect(promise).rejects.toThrow('EBUSY');
      expect(operation).toHaveBeenCalledTimes(3); // Initial + 2 retries
    });

    it('does not retry on non-transient error', async () => {
      const permanentError = new Error('Permanent failure');
      const operation = vi.fn().mockRejectedValue(permanentError);

      await expect(
        withRetry(operation, { maxRetries: 3 })
      ).rejects.toThrow('Permanent failure');

      expect(operation).toHaveBeenCalledTimes(1); // No retries
    });

    it('calls onRetry callback on each retry', async () => {
      const transientError = Object.assign(new Error('EBUSY'), { code: 'EBUSY' });
      const operation = vi.fn()
        .mockRejectedValueOnce(transientError)
        .mockResolvedValueOnce('success');
      const onRetry = vi.fn();

      const promise = withRetry(operation, {
        maxRetries: 2,
        baseDelay: 100,
        onRetry,
      });

      await vi.advanceTimersByTimeAsync(100);
      await promise;

      expect(onRetry).toHaveBeenCalledTimes(1);
      expect(onRetry).toHaveBeenCalledWith(transientError, 1, 100);
    });

    it('uses exponential backoff by default', async () => {
      const transientError = Object.assign(new Error('EBUSY'), { code: 'EBUSY' });
      const operation = vi.fn().mockRejectedValue(transientError);
      const onRetry = vi.fn();

      const promise = withRetry(operation, {
        maxRetries: 3,
        baseDelay: 100,
        onRetry,
      });

      await vi.advanceTimersByTimeAsync(100);  // 1st retry: 100ms
      await vi.advanceTimersByTimeAsync(200);  // 2nd retry: 200ms
      await vi.advanceTimersByTimeAsync(400);  // 3rd retry: 400ms

      await expect(promise).rejects.toThrow();

      expect(onRetry).toHaveBeenNthCalledWith(1, transientError, 1, 100);
      expect(onRetry).toHaveBeenNthCalledWith(2, transientError, 2, 200);
      expect(onRetry).toHaveBeenNthCalledWith(3, transientError, 3, 400);
    });

    it('respects maxDelay cap', async () => {
      const transientError = Object.assign(new Error('EBUSY'), { code: 'EBUSY' });
      const operation = vi.fn().mockRejectedValue(transientError);
      const onRetry = vi.fn();

      const promise = withRetry(operation, {
        maxRetries: 10,
        baseDelay: 100,
        maxDelay: 500,
        onRetry,
      });

      // Exponential would be: 100, 200, 400, 800, 1600...
      // With maxDelay=500: 100, 200, 400, 500, 500...
      await vi.advanceTimersByTimeAsync(100);
      await vi.advanceTimersByTimeAsync(200);
      await vi.advanceTimersByTimeAsync(400);
      await vi.advanceTimersByTimeAsync(500);  // Capped at 500
      await vi.advanceTimersByTimeAsync(500);  // Still capped

      await expect(promise).rejects.toThrow();

      const delays = onRetry.mock.calls.map(call => call[2]);
      expect(delays.slice(3)).toEqual([500, 500, 500, 500, 500, 500, 500]); // All capped
    });
  });

  describe('createErrorNotification', () => {
    it('creates error notification from Error', () => {
      const error = new Error('Something went wrong');
      const notification = createErrorNotification(error);

      expect(notification.type).toBe('error');
      expect(notification.message).toBe('Something went wrong');
    });

    it('creates warning notification for permission errors', () => {
      const error = Object.assign(new Error('EACCES'), { code: 'EACCES' });
      const notification = createErrorNotification(error);

      expect(notification.type).toBe('warning');
    });

    it('creates warning notification for not found errors', () => {
      const error = Object.assign(new Error('ENOENT'), { code: 'ENOENT' });
      const notification = createErrorNotification(error);

      expect(notification.type).toBe('warning');
    });

    it('uses default message for empty error message', () => {
      const error = new Error('');
      const notification = createErrorNotification(error, 'Default message');

      expect(notification.message).toBe('Default message');
    });

    it('handles non-Error objects', () => {
      const notification = createErrorNotification('string error', 'Default');

      expect(notification.message).toBe('string error');
    });
  });

  describe('isTransientError', () => {
    it('returns true for EBUSY', () => {
      const error = Object.assign(new Error('EBUSY'), { code: 'EBUSY' });
      expect(isTransientError(error)).toBe(true);
    });

    it('returns true for EAGAIN', () => {
      const error = Object.assign(new Error('EAGAIN'), { code: 'EAGAIN' });
      expect(isTransientError(error)).toBe(true);
    });

    it('returns false for permanent errors', () => {
      const error = Object.assign(new Error('ENOENT'), { code: 'ENOENT' });
      expect(isTransientError(error)).toBe(false);
    });

    it('returns false for regular errors', () => {
      const error = new Error('Regular error');
      expect(isTransientError(error)).toBe(false);
    });
  });
});
```

Create `tests/utils/test_errorTypes.ts`:

```typescript
import { describe, it, expect } from 'vitest';
import {
  GitError,
  FileSystemError,
  ConfigError,
  isYellowwoodError,
  isPermissionError,
  isNotFoundError,
  getUserMessage,
  getErrorDetails,
} from '../../src/utils/errorTypes.js';

describe('errorTypes', () => {
  describe('YellowwoodError classes', () => {
    it('creates GitError with message and context', () => {
      const error = new GitError('Git failed', { repo: '/path/to/repo' });

      expect(error.message).toBe('Git failed');
      expect(error.context).toEqual({ repo: '/path/to/repo' });
      expect(error.name).toBe('GitError');
    });

    it('chains errors with cause', () => {
      const cause = new Error('Original error');
      const error = new GitError('Git failed', {}, cause);

      expect(error.cause).toBe(cause);
    });

    it('creates FileSystemError', () => {
      const error = new FileSystemError('Permission denied', { path: '/root' });

      expect(error.message).toBe('Permission denied');
      expect(error.name).toBe('FileSystemError');
    });

    it('creates ConfigError', () => {
      const error = new ConfigError('Invalid config');

      expect(error.message).toBe('Invalid config');
      expect(error.name).toBe('ConfigError');
    });
  });

  describe('isYellowwoodError', () => {
    it('returns true for YellowwoodError subclasses', () => {
      expect(isYellowwoodError(new GitError('test'))).toBe(true);
      expect(isYellowwoodError(new FileSystemError('test'))).toBe(true);
      expect(isYellowwoodError(new ConfigError('test'))).toBe(true);
    });

    it('returns false for regular Error', () => {
      expect(isYellowwoodError(new Error('test'))).toBe(false);
    });

    it('returns false for non-errors', () => {
      expect(isYellowwoodError('string')).toBe(false);
      expect(isYellowwoodError(null)).toBe(false);
      expect(isYellowwoodError(undefined)).toBe(false);
    });
  });

  describe('isPermissionError', () => {
    it('returns true for EACCES', () => {
      const error = Object.assign(new Error('EACCES'), { code: 'EACCES' });
      expect(isPermissionError(error)).toBe(true);
    });

    it('returns true for EPERM', () => {
      const error = Object.assign(new Error('EPERM'), { code: 'EPERM' });
      expect(isPermissionError(error)).toBe(true);
    });

    it('returns false for other error codes', () => {
      const error = Object.assign(new Error('ENOENT'), { code: 'ENOENT' });
      expect(isPermissionError(error)).toBe(false);
    });
  });

  describe('isNotFoundError', () => {
    it('returns true for ENOENT', () => {
      const error = Object.assign(new Error('ENOENT'), { code: 'ENOENT' });
      expect(isNotFoundError(error)).toBe(true);
    });

    it('returns false for other error codes', () => {
      const error = Object.assign(new Error('EACCES'), { code: 'EACCES' });
      expect(isNotFoundError(error)).toBe(false);
    });
  });

  describe('getUserMessage', () => {
    it('extracts message from YellowwoodError', () => {
      const error = new GitError('Git failed');
      expect(getUserMessage(error)).toBe('Git failed');
    });

    it('extracts message from Error', () => {
      const error = new Error('Generic error');
      expect(getUserMessage(error)).toBe('Generic error');
    });

    it('converts non-errors to string', () => {
      expect(getUserMessage('string error')).toBe('string error');
      expect(getUserMessage(123)).toBe('123');
    });
  });

  describe('getErrorDetails', () => {
    it('extracts details from YellowwoodError', () => {
      const cause = new Error('Cause');
      const error = new GitError('Git failed', { repo: '/path' }, cause);
      const details = getErrorDetails(error);

      expect(details.message).toBe('Git failed');
      expect(details.name).toBe('GitError');
      expect(details.context).toEqual({ repo: '/path' });
      expect(details.cause).toBeDefined();
    });

    it('extracts Node.js error codes', () => {
      const error = Object.assign(new Error('ENOENT'), {
        code: 'ENOENT',
        syscall: 'open',
        path: '/missing',
      });
      const details = getErrorDetails(error);

      expect(details.code).toBe('ENOENT');
      expect(details.syscall).toBe('open');
      expect(details.path).toBe('/missing');
    });
  });
});
```

## Technical Specifications

**Footprint:**
- Modules touched: error-handling, git-integration, config-system, file-watching, app-orchestration
- Files created: 
  - `src/utils/errorHandling.ts` (main error handling utilities)
  - `src/utils/errorTypes.ts` (custom error classes)
  - `src/utils/logger.ts` (structured logging)
  - `tests/utils/test_errorHandling.ts` (error handling tests)
  - `tests/utils/test_errorTypes.ts` (error types tests)
- Files modified:
  - `src/utils/git.ts` (add error boundaries)
  - `src/utils/config.ts` (improve error messages)
  - `src/App.tsx` (integrate global error handler)
  - Other utilities as needed

**Dependencies:**
- No new external dependencies required
- Uses existing types from `src/types/index.ts`

**Size Estimate:**
- Size: M (3 points)
- Effort: 8-10 hours total
  - Error types and logger: 1.5 hours
  - Error boundaries and retry: 3 hours
  - Integration with existing code: 2 hours
  - Comprehensive testing: 3 hours
  - Documentation and examples: 0.5 hours

## Dependencies

**Strict Dependencies (Blocking):**
- Issue #2 (Notification system) - Must exist to show error notifications

**Flexible Dependencies (Informational):**
- Issue #1 (Configuration loading) - Will enhance with error handling
- Issue #3 (Git status detection) - Will add error boundaries
- Issue #11 (File watcher) - Will add retry logic for watcher setup

**Assumptions:**
- Notification component exists and works (#2)
- App.tsx manages notification state
- TypeScript strict mode is enabled

## Acceptance Criteria

- [ ] Custom error types defined (GitError, FileSystemError, ConfigError, ProcessError, WatcherError)
- [ ] Error boundary functions implemented (withGitErrorBoundary, withFileSystemErrorBoundary, withConfigErrorBoundary)
- [ ] Retry logic with exponential backoff implemented (withRetry)
- [ ] Structured logging utilities created (logDebug, logInfo, logWarn, logError)
- [ ] Global error handler setup function implemented
- [ ] Permission errors show user-friendly "Permission denied: path" messages
- [ ] File not found errors show "File not found: path" messages
- [ ] Git errors fall back gracefully (disable git features, show info notification)
- [ ] Config errors fall back to DEFAULT_CONFIG with warning notification
- [ ] Transient errors (EBUSY, EAGAIN) are retried automatically
- [ ] Error notifications have appropriate severity (error/warning/info)
- [ ] All error paths are logged with appropriate severity
- [ ] Error context includes relevant debugging information
- [ ] Unit tests cover all error scenarios (20+ test cases)
- [ ] Integration tests verify error recovery in App.tsx
- [ ] Manual testing confirms helpful error messages

## Implementation Guidance

**Recommended Approach:**

1. **Create error type system** (1.5 hours)
   - Define custom error classes
   - Add type guards and helpers
   - Test error creation and chaining

2. **Implement logger** (30 min)
   - Simple console-based logging
   - Structured output format
   - Test different log levels

3. **Build error boundaries** (2 hours)
   - Start with withGitErrorBoundary
   - Add withFileSystemErrorBoundary with permission/notfound handling
   - Add withConfigErrorBoundary
   - Test each boundary independently

4. **Implement retry logic** (1.5 hours)
   - Start with simple retry
   - Add exponential backoff
   - Add maxDelay cap
   - Test with fake timers

5. **Integrate with existing code** (2 hours)
   - Update git.ts to throw custom errors
   - Update config.ts with better messages
   - Add error boundaries to App.tsx
   - Test end-to-end error flows

6. **Write comprehensive tests** (3 hours)
   - Test error boundaries with various errors
   - Test retry logic with transient errors
   - Test error notification creation
   - Test global error handler setup

7. **Manual verification** (1 hour)
   - Trigger permission errors
   - Test with missing git
   - Test with invalid config
   - Verify all error messages are helpful

**Patterns to Follow:**

- **Fail gracefully**: Never crash the app - always provide fallback
- **Be specific**: Use custom error types to distinguish error categories
- **Log appropriately**: Debug for development, warn for expected issues, error for unexpected
- **User-first messages**: Errors should tell users what went wrong and optionally how to fix it
- **Include context**: Error objects should carry debugging information
- **Clean up resources**: Always return cleanup functions from useEffect

**Patterns to Avoid:**

- Don't swallow errors silently (always log)
- Don't show technical stack traces to users
- Don't retry non-transient errors (wastes time)
- Don't use generic "An error occurred" messages when specific context is available
- Don't let one subsystem failure crash the entire app

**Code Quality Checklist:**

- [ ] All error types extend YellowwoodError
- [ ] All async functions have try/catch or use error boundaries
- [ ] All error messages are user-friendly
- [ ] All errors include context for debugging
- [ ] Retry logic has reasonable limits (max retries, max delay)
- [ ] Global error handler cleans up on unmount
- [ ] TypeScript strict mode passes
- [ ] No console.log (only structured logging)

## Edge Cases & Gotchas

**Node.js vs Browser:**
- Global error handlers differ between Node and browser
- Check for `window` existence before using browser APIs
- Use `process.on()` for Node.js uncaught errors

**Error chaining:**
- Always preserve original error as `cause`
- Don't lose stack traces when wrapping errors
- Use `Error.captureStackTrace()` where available

**Retry infinite loops:**
- Always have a max retry limit
- Always have a max delay cap
- Check shouldRetry predicate to avoid retrying permanent errors

**Memory leaks:**
- Clear timers in retry logic
- Remove event listeners in global error handler cleanup
- Don't accumulate error logs (they're just console output)

**Notification spam:**
- Don't show notification for every retry attempt
- Only notify on final failure
- Auto-dismiss non-errors to avoid cluttering UI

**Permission errors on Windows:**
- Windows uses different error codes than Unix
- Test on multiple platforms
- EPERM vs EACCES may behave differently

**Git subprocess errors:**
- Git commands can fail in many ways (not installed, network, corruption)
- Distinguish between "no git" and "git failed"
- "No git" is info, "git failed" is error

**Config validation timing:**
- Validate early (on load) not late (on use)
- Invalid config should prevent startup with helpful message
- Missing config is fine (use defaults)

**Async error boundaries:**
- Try/catch only works for sync code or awaited promises
- Fire-and-forget promises need global handlers
- Always await async operations in try/catch

**Error serialization:**
- Errors don't serialize to JSON well (lose stack traces)
- Use getErrorDetails() for logging
- Don't try to send errors over IPC/network without serialization

## Resources

**External Documentation:**
- [Error Handling in Node.js](https://nodejs.org/api/errors.html) - Node.js error types
- [TypeScript Error Handling](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) - Type guards
- [Exponential Backoff](https://en.wikipedia.org/wiki/Exponential_backoff) - Retry strategy
- [React Error Boundaries](https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary) - Error handling in React

**Internal References:**
- `src/types/index.ts:19-22` - Notification interface
- `src/types/index.ts:5` - NotificationType
- `src/utils/git.ts` - Git operations that need error boundaries
- `src/utils/config.ts` - Config loading that needs error handling
- SPEC.md section 10 "Error Handling" - Requirements and strategy

**Similar Patterns:**
- VSCode error handling architecture
- ESLint error recovery strategies
- Prettier graceful degradation

## Notes

**Why this matters:**
Yellowwood interacts with many external systems (git, filesystem, processes). Without robust error handling, any of these could crash the app or leave it in a broken state. This issue ensures the app is resilient and provides helpful feedback.

**Design philosophy:**
- **User-centric**: Errors should help users understand what's wrong and what to do
- **Defensive**: Assume everything can fail - wrap accordingly
- **Graceful**: Degraded functionality is better than crashed app
- **Transparent**: Log everything for debugging, show users only what's helpful

**Testing strategy:**
- Unit test each error boundary in isolation
- Integration test error recovery in App.tsx
- Manual test by triggering real errors (permissions, missing files, etc.)
- Use fake timers for retry logic tests
- Mock filesystem and git for controlled error scenarios

**Future enhancements (NOT in this issue):**
- Error reporting/telemetry (opt-in)
- Error recovery suggestions ("Try running 'git init'")
- Error history panel (view recent errors)
- Custom error handlers per operation type
- Circuit breaker pattern for repeatedly failing operations
- Rate limiting for error notifications
- Sentry/Bugsnag integration for production monitoring

**Integration notes:**
This error handling system will be used throughout Yellowwood:
- All utilities (git, config, file watcher) wrap operations in error boundaries
- All hooks (useGitStatus, useFileTree) handle errors gracefully
- App.tsx sets up global error handler on mount
- Notifications provide user feedback for all errors
- Logging provides developer insight for debugging

**Performance considerations:**
- Retry logic adds latency (by design)
- Logging is synchronous - keep minimal
- Error object creation is cheap
- Don't retry fast - use exponential backoff
- Cap retry attempts and delays to avoid hanging

**Compatibility notes:**
- Error codes (EACCES, ENOENT, etc.) are cross-platform
- Process error handlers work in Node.js CLI apps
- Browser error handlers are for future web version
- Windows, macOS, and Linux all supported

--- a/src/utils/config.ts
+++ b/src/utils/config.ts
@@ -3,6 +3,8 @@ import os from 'os';
 import path from 'path';
 import fs from 'fs-extra';
 import { DEFAULT_CONFIG, type YellowwoodConfig } from '../types/index.js';
+import { ConfigError } from './errorTypes.js';
+import { logWarn, logError } from './logger.js';
 
 /**
  * Loads Yellowwood configuration from project and global config files.
@@ -31,7 +33,7 @@ export async function loadConfig(cwd: string = process.cwd()): Promise<Yellowwoo
     projectConfig = projectResult?.config || {};
   } catch (error) {
     // Handle malformed JSON or other config file errors
-    console.warn(`Warning: Could not load project config:`, (error as Error).message);
+    logWarn('Could not load project config', { cwd, error: (error as Error).message });
     // Continue with empty project config (fall back to global/defaults)
   }
 
@@ -69,7 +71,7 @@ async function loadGlobalConfig(): Promise<Partial<YellowwoodConfig>> {
       return {}; // File disappeared between check and read - that's OK
     }
     // JSON parse error or permission error - log and continue with empty config
-    console.warn(`Warning: Could not load global config from ${globalPath}:`, (error as Error).message);
+    logWarn('Could not load global config', { globalPath, error: (error as Error).message });
     return {};
   }
 }
@@ -330,14 +332,17 @@ function validateConfig(config: unknown): YellowwoodConfig {
   }
 
   if (errors.length > 0) {
-    throw new Error(`Config validation failed:\n${errors.join('\n')}`);
+    throw new ConfigError(
+      `Config validation failed:\n${errors.join('\n')}`,
+      { config: c }
+    );
   }
 
   // Warn about unknown keys
   const knownKeys = Object.keys(DEFAULT_CONFIG);
   const unknownKeys = Object.keys(c).filter(key => !knownKeys.includes(key));
   if (unknownKeys.length > 0) {
-    console.warn(`Warning: Unknown config keys (will be ignored): ${unknownKeys.join(', ')}`);
+    logWarn('Unknown config keys (will be ignored)', { keys: unknownKeys });
   }
 
   return c as YellowwoodConfig;

--- a/src/utils/errorHandling.ts
+++ b/src/utils/errorHandling.ts
@@ -1,0 +1,290 @@
+/**
+ * Error boundary and retry utilities for robust error handling.
+ * Provides graceful degradation and automatic recovery strategies.
+ */
+
+import type { Notification, NotificationType } from '../types/index.js';
+import { logError, logWarn, logDebug } from './logger.js';
+import {
+  GitError,
+  FileSystemError,
+  ConfigError,
+  ProcessError,
+  WatcherError,
+  isTransientError,
+  isPermissionError,
+  isNotFoundError,
+  getUserMessage,
+  getErrorDetails,
+} from './errorTypes.js';
+
+/**
+ * Configuration for error boundaries
+ */
+interface ErrorBoundaryOptions<T> {
+  fallback: T;                          // Value to return on error
+  notify?: boolean;                     // Show notification to user
+  notificationType?: NotificationType;  // Override notification type
+  onError?: (error: unknown) => void;   // Custom error handler
+  context?: Record<string, unknown>;    // Additional context for logging
+}
+
+/**
+ * Configuration for retry logic
+ */
+interface RetryOptions {
+  maxRetries?: number;           // Maximum number of retries (default: 3)
+  backoff?: 'none' | 'linear' | 'exponential';  // Backoff strategy (default: exponential)
+  baseDelay?: number;            // Base delay in ms (default: 100)
+  maxDelay?: number;             // Maximum delay in ms (default: 5000)
+  shouldRetry?: (error: unknown, attempt: number) => boolean;  // Custom retry predicate
+  notify?: boolean;              // Show notification on final failure
+  onRetry?: (error: unknown, attempt: number, delay: number) => void;  // Retry callback
+}
+
+/**
+ * Wrap a git operation in an error boundary.
+ * Catches git errors and provides fallback value.
+ */
+export async function withGitErrorBoundary<T>(
+  operation: () => Promise<T>,
+  options: ErrorBoundaryOptions<T>
+): Promise<T> {
+  try {
+    return await operation();
+  } catch (error) {
+    const gitError = error instanceof GitError ? error : new GitError(
+      'Git operation failed',
+      options.context,
+      error instanceof Error ? error : undefined
+    );
+
+    logError('Git error', gitError, options.context);
+
+    if (options.notify && options.onError) {
+      try {
+        options.onError(gitError);
+      } catch (hookError) {
+        logWarn('Error boundary notification handler failed', { hookError });
+      }
+    }
+
+    return options.fallback;
+  }
+}
+
+/**
+ * Wrap a file system operation in an error boundary.
+ * Handles permission errors, not found errors, and other filesystem failures.
+ */
+export async function withFileSystemErrorBoundary<T>(
+  operation: () => Promise<T>,
+  options: ErrorBoundaryOptions<T>
+): Promise<T> {
+  try {
+    return await operation();
+  } catch (error) {
+    let fsError: FileSystemError;
+    let userMessage = 'File system operation failed';
+
+    if (isPermissionError(error)) {
+      const path = (error as NodeJS.ErrnoException).path || 'unknown';
+      userMessage = `Permission denied: Cannot access ${path}`;
+    } else if (isNotFoundError(error)) {
+      const path = (error as NodeJS.ErrnoException).path || 'unknown';
+      userMessage = `File not found: ${path}`;
+    }
+
+    fsError = error instanceof FileSystemError ? error : new FileSystemError(
+      userMessage,
+      options.context,
+      error instanceof Error ? error : undefined
+    );
+
+    const severity = isPermissionError(error) || isNotFoundError(error) ? 'warn' : 'error';
+    if (severity === 'warn') {
+      logWarn(userMessage, options.context);
+    } else {
+      logError(userMessage, fsError, options.context);
+    }
+
+    if (options.notify && options.onError) {
+      try {
+        options.onError(fsError);
+      } catch (hookError) {
+        logWarn('Error boundary notification handler failed', { hookError });
+      }
+    }
+
+    return options.fallback;
+  }
+}
+
+/**
+ * Wrap a config operation in an error boundary.
+ * Validates and provides helpful messages for config errors.
+ */
+export async function withConfigErrorBoundary<T>(
+  operation: () => Promise<T>,
+  options: ErrorBoundaryOptions<T>
+): Promise<T> {
+  try {
+    return await operation();
+  } catch (error) {
+    const configError = error instanceof ConfigError ? error : new ConfigError(
+      'Configuration error',
+      options.context,
+      error instanceof Error ? error : undefined
+    );
+
+    logError('Config error', configError, options.context);
+
+    if (options.notify && options.onError) {
+      try {
+        options.onError(configError);
+      } catch (hookError) {
+        logWarn('Error boundary notification handler failed', { hookError });
+      }
+    }
+
+    return options.fallback;
+  }
+}
+
+/**
+ * Retry an operation with exponential backoff.
+ * Useful for transient errors like file locks or network issues.
+ */
+export async function withRetry<T>(
+  operation: () => Promise<T>,
+  options: RetryOptions = {}
+): Promise<T> {
+  const {
+    maxRetries = 3,
+    backoff = 'exponential',
+    baseDelay = 100,
+    maxDelay = 5000,
+    shouldRetry = isTransientError,
+    notify = true,
+    onRetry,
+  } = options;
+
+  let lastError: unknown;
+  let attempt = 0;
+
+  while (attempt <= maxRetries) {
+    try {
+      return await operation();
+    } catch (error) {
+      lastError = error;
+      attempt++;
+
+      // Check if we should retry
+      if (attempt > maxRetries || !shouldRetry(error, attempt)) {
+        break;
+      }
+
+      // Calculate delay
+      let delay = baseDelay;
+      if (backoff === 'linear') {
+        delay = baseDelay * attempt;
+      } else if (backoff === 'exponential') {
+        delay = Math.min(baseDelay * Math.pow(2, attempt - 1), maxDelay);
+      }
+
+      logDebug(`Retrying operation (attempt ${attempt}/${maxRetries}) after ${delay}ms`, {
+        error: getUserMessage(error),
+      });
+
+      onRetry?.(error, attempt, delay);
+
+      // Wait before retrying
+      await new Promise(resolve => setTimeout(resolve, delay));
+    }
+  }
+
+  // All retries exhausted
+  const message = `Operation failed after ${attempt} attempts`;
+  logError(message, lastError);
+
+  if (notify) {
+    // Caller should handle notification via thrown error
+  }
+
+  throw lastError;
+}
+
+/**
+ * Safely execute an operation with comprehensive error handling.
+ * Combines error boundary + retry logic + logging.
+ */
+export async function safeExecute<T>(
+  operation: () => Promise<T>,
+  options: {
+    fallback: T;
+    retry?: RetryOptions;
+    boundary?: Omit<ErrorBoundaryOptions<T>, 'fallback'>;
+  }
+): Promise<T> {
+  const boundaryOptions: ErrorBoundaryOptions<T> = {
+    fallback: options.fallback,
+    ...options.boundary,
+  };
+
+  return withFileSystemErrorBoundary(async () => {
+    if (options.retry) {
+      return await withRetry(operation, options.retry);
+    }
+    return await operation();
+  }, boundaryOptions);
+}
+
+/**
+ * Create a user-friendly notification from an error.
+ */
+export function createErrorNotification(
+  error: unknown,
+  defaultMessage: string = 'An error occurred'
+): Notification {
+  let message = getUserMessage(error);
+  let type: NotificationType = 'error';
+
+  // Downgrade severity for certain error types
+  if (isPermissionError(error) || isNotFoundError(error)) {
+    type = 'warning';
+  }
+
+  // If message is too technical, use default
+  if (!message || message.length < 10) {
+    message = defaultMessage;
+  }
+
+  return { message, type };
+}
+
+/**
+ * Handle uncaught errors globally.
+ * Should be set up in App.tsx on mount.
+ */
+export function setupGlobalErrorHandler(
+  onError: (error: unknown) => void
+): () => void {
+  // For Node.js CLI apps
+  const processErrorHandler = (error: Error) => {
+    logError('Uncaught exception', error);
+    onError(error);
+  };
+
+  const processRejectionHandler = (reason: unknown) => {
+    logError('Unhandled rejection', reason);
+    onError(reason);
+  };
+
+  process.on('uncaughtException', processErrorHandler);
+  process.on('unhandledRejection', processRejectionHandler);
+
+  return () => {
+    process.off('uncaughtException', processErrorHandler);
+    process.off('unhandledRejection', processRejectionHandler);
+  };
+}

--- a/src/utils/errorTypes.ts
+++ b/src/utils/errorTypes.ts
@@ -1,0 +1,153 @@
+/**
+ * Custom error types for Yellowwood with context and severity.
+ * Enables better error handling, logging, and user notifications.
+ */
+
+/**
+ * Base error class for all Yellowwood errors.
+ * Includes context for better debugging and user messages.
+ */
+export class YellowwoodError extends Error {
+  constructor(
+    message: string,
+    public readonly context?: Record<string, unknown>,
+    public readonly cause?: Error
+  ) {
+    super(message);
+    this.name = this.constructor.name;
+    Error.captureStackTrace?.(this, this.constructor);
+  }
+}
+
+/**
+ * Git operation failed (repository not found, git not installed, command failed)
+ */
+export class GitError extends YellowwoodError {
+  constructor(message: string, context?: Record<string, unknown>, cause?: Error) {
+    super(message, context, cause);
+  }
+}
+
+/**
+ * File system operation failed (permission denied, file not found, read error)
+ */
+export class FileSystemError extends YellowwoodError {
+  constructor(message: string, context?: Record<string, unknown>, cause?: Error) {
+    super(message, context, cause);
+  }
+}
+
+/**
+ * Configuration loading or validation failed
+ */
+export class ConfigError extends YellowwoodError {
+  constructor(message: string, context?: Record<string, unknown>, cause?: Error) {
+    super(message, context, cause);
+  }
+}
+
+/**
+ * External process execution failed (editor not found, command error)
+ */
+export class ProcessError extends YellowwoodError {
+  constructor(message: string, context?: Record<string, unknown>, cause?: Error) {
+    super(message, context, cause);
+  }
+}
+
+/**
+ * File watcher setup or operation failed
+ */
+export class WatcherError extends YellowwoodError {
+  constructor(message: string, context?: Record<string, unknown>, cause?: Error) {
+    super(message, context, cause);
+  }
+}
+
+/**
+ * Check if error is a specific Yellowwood error type
+ */
+export function isYellowwoodError(error: unknown): error is YellowwoodError {
+  return error instanceof YellowwoodError;
+}
+
+/**
+ * Check if error is a permission/access error (EACCES, EPERM)
+ */
+export function isPermissionError(error: unknown): boolean {
+  if (!error || typeof error !== 'object') return false;
+  const code = (error as NodeJS.ErrnoException).code;
+  return code === 'EACCES' || code === 'EPERM';
+}
+
+/**
+ * Check if error is a "not found" error (ENOENT)
+ */
+export function isNotFoundError(error: unknown): boolean {
+  if (!error || typeof error !== 'object') return false;
+  return (error as NodeJS.ErrnoException).code === 'ENOENT';
+}
+
+/**
+ * Check if error is a transient error that might succeed on retry
+ */
+export function isTransientError(error: unknown): boolean {
+  if (!error || typeof error !== 'object') return false;
+  const code = (error as NodeJS.ErrnoException).code;
+  // File busy, resource temporarily unavailable, etc.
+  return ['EBUSY', 'EAGAIN', 'ETIMEDOUT', 'ECONNRESET', 'ENOTFOUND'].includes(code || '');
+}
+
+/**
+ * Extract user-friendly message from any error
+ */
+export function getUserMessage(error: unknown): string {
+  if (isYellowwoodError(error)) {
+    return error.message;
+  }
+
+  if (error instanceof Error) {
+    return error.message;
+  }
+
+  return String(error);
+}
+
+/**
+ * Extract technical details for logging
+ * Handles circular references safely to prevent infinite recursion
+ */
+export function getErrorDetails(error: unknown, seen = new WeakSet<Error>()): Record<string, unknown> {
+  const details: Record<string, unknown> = {
+    message: getUserMessage(error),
+  };
+
+  if (error instanceof Error) {
+    details.name = error.name;
+    details.stack = error.stack;
+  }
+
+  if (isYellowwoodError(error)) {
+    details.context = error.context;
+    if (error.cause) {
+      // Prevent circular reference stack overflow
+      if (error.cause instanceof Error && !seen.has(error.cause)) {
+        seen.add(error.cause);
+        details.cause = getErrorDetails(error.cause, seen);
+      } else if (!(error.cause instanceof Error)) {
+        // Handle non-Error causes
+        details.cause = getErrorDetails(error.cause, seen);
+      }
+    }
+  }
+
+  if (error && typeof error === 'object') {
+    const nodeError = error as NodeJS.ErrnoException;
+    if (nodeError.code) details.code = nodeError.code;
+    if (nodeError.errno) details.errno = nodeError.errno;
+    if (nodeError.syscall) details.syscall = nodeError.syscall;
+    if (nodeError.path) details.path = nodeError.path;
+  }
+
+  return details;
+}

--- a/src/utils/fileWatcher.ts
+++ b/src/utils/fileWatcher.ts
@@ -1,5 +1,7 @@
 import chokidar, { FSWatcher, ChokidarOptions } from 'chokidar';
 import path from 'path';
+import { WatcherError } from './errorTypes.js';
+import { logWarn, logError, logInfo } from './logger.js';
 
 export interface FileWatcherOptions {
 	ignored?: ChokidarOptions['ignored']; // Patterns to ignore (supports string, RegExp, function, etc.)
@@ -100,7 +102,7 @@ export function createFileWatcher(
 
 	const start = (): void => {
 		if (isActive) {
-			console.warn('File watcher is already running');
+			logWarn('File watcher is already running', { rootPath });
 			return;
 		}
 
@@ -172,7 +174,8 @@ export function createFileWatcher(
 			// Error handler
 			watcher.on('error', (error: unknown) => {
 				const err = error instanceof Error ? error : new Error(String(error));
-				console.error('File watcher error:', err.message);
+				const watcherError = new WatcherError('File watcher error', { rootPath }, err);
+				logError('File watcher error', watcherError);
 
 				// Stop watcher on fatal errors (ENOSPC, EPERM, etc.)
 				// This allows the caller to detect the failure and handle recovery
@@ -182,29 +185,35 @@ export function createFileWatcher(
 					// and allow clean restart if needed
 					if (watcher) {
 						watcher.close().catch((closeError) => {
-							console.error('Error closing watcher after error:', closeError);
+							logError('Error closing watcher after error', closeError);
 						});
 						watcher = null;
 					}
 				}
 
 				if (onError) {
-					onError(err);
+					onError(watcherError);
 				}
 			});
 
 			// Ready handler (watcher is initialized)
 			watcher.on('ready', () => {
-				console.log(`File watcher started for: ${rootPath}`);
+				logInfo('File watcher started', { rootPath });
 			});
 
 			isActive = true;
 		} catch (error) {
-			console.error('Failed to start file watcher:', (error as Error).message);
+			const watcherError = new WatcherError(
+		// Normalize error cause to always be an Error instance
+				'Failed to start file watcher',
+				{ rootPath },
+				error instanceof Error ? error : new Error(String(error))
+			);
+			logError('Failed to start file watcher', watcherError);
 			if (onError) {
-				onError(error as Error);
+				onError(watcherError);
 			}
-			throw error;
+			throw watcherError;
 		}
 	};
 
@@ -223,16 +232,22 @@ export function createFileWatcher(
 		// Close the watcher
 		try {
 			await watcher.close();
-			console.log('File watcher stopped');
+			logInfo('File watcher stopped', { rootPath });
 			watcher = null;
 			isActive = false;
 		} catch (error) {
-			console.error('Error stopping file watcher:', (error as Error).message);
+		// Normalize error cause to always be an Error instance
+			const watcherError = new WatcherError(
+				'Error stopping file watcher',
+				{ rootPath },
+				error instanceof Error ? error : undefined
+			);
+			logError('Error stopping file watcher', watcherError);
 			// Still mark as stopped even if close fails
 			watcher = null;
 			isActive = false;
 			// Rethrow so caller knows about the failure
-			throw error;
+			throw watcherError;
 		}
 	};
 

--- a/src/utils/git.test.ts
+++ b/src/utils/git.test.ts
@@ -4,6 +4,7 @@ import fs from 'fs-extra';
 import path from 'path';
 import os from 'os';
 import { simpleGit } from 'simple-git';
+import { GitError } from './errorTypes.js';
 
 describe('git utilities', () => {
   let testRepoPath: string;
@@ -140,9 +141,17 @@ describe('git utilities', () => {
       expect(statusMap.size).toBe(3);
     });
 
-    it('returns empty map for non-git directory', async () => {
-      const statusMap = await getGitStatus(nonRepoPath);
-      expect(statusMap.size).toBe(0);
+    it('throws GitError for non-git directory', async () => {
+      await expect(getGitStatus(nonRepoPath)).rejects.toThrow(GitError);
+      await expect(getGitStatus(nonRepoPath)).rejects.toThrow('Failed to get git status');
+
+      // Verify error context includes cwd
+      try {
+        await getGitStatus(nonRepoPath);
+      } catch (error) {
+        expect(error).toBeInstanceOf(GitError);
+        expect((error as GitError).context?.cwd).toBe(nonRepoPath);
+      }
     });
 
     it('returns absolute paths in the map', async () => {
@@ -262,9 +271,17 @@ describe('git utilities', () => {
       const isRepo = await isGitRepository(nonexistentPath);
       expect(isRepo).toBe(false);
 
-      // getGitStatus should also handle gracefully
-      const statusMap = await getGitStatus(nonexistentPath);
-      expect(statusMap.size).toBe(0);
+      // getGitStatus should throw GitError for nonexistent directory
+      await expect(getGitStatus(nonexistentPath)).rejects.toThrow(GitError);
+      await expect(getGitStatus(nonexistentPath)).rejects.toThrow('Failed to get git status');
+
+      // Verify error context
+      try {
+        await getGitStatus(nonexistentPath);
+      } catch (error) {
+        expect(error).toBeInstanceOf(GitError);
+        expect((error as GitError).context?.cwd).toBe(nonexistentPath);
+      }
     });
 
     it('excludes ignored files from status', async () => {

--- a/src/utils/git.ts
+++ b/src/utils/git.ts
@@ -2,6 +2,8 @@ import { resolve } from 'path';
 import { realpathSync } from 'fs';
 import simpleGit, { SimpleGit, StatusResult } from 'simple-git';
 import type { GitStatus } from '../types/index.js';
+import { GitError } from './errorTypes.js';
+import { logWarn, logError } from './logger.js';
 
 /**
  * Check if a directory is inside a git repository.
@@ -15,7 +17,8 @@ export async function isGitRepository(path: string): Promise<boolean> {
     const isRepo = await git.checkIsRepo();
     return isRepo;
   } catch (error) {
-    // Git not installed or other error
+    // Git not installed or other error - this is expected, not an error
+    logWarn('Git repository check failed', { path, error: (error as Error).message });
     return false;
   }
 }
@@ -92,9 +95,20 @@ export async function getGitStatus(cwd: string): Promise<Map<string, GitStatus>>
     // - Can be added in a future enhancement if needed
 
   } catch (error) {
-    // Not a git repo or git command failed
-    // Return empty map (no git status available)
-    console.warn('Failed to get git status:', (error as Error).message);
+    // Normalize error cause to always be an Error instance
+    const cause = error instanceof Error ? error : new Error(String(error));
+
+    // Git operation failed - wrap in GitError for better context
+    const gitError = new GitError(
+      'Failed to get git status',
+      { cwd },
+      cause
+    );
+
+    // Log before throwing so failures appear in structured logs
+    logError('Git status operation failed', gitError, { cwd });
+
+    throw gitError;
   }
 
   return statusMap;

--- a/src/utils/logger.ts
+++ b/src/utils/logger.ts
@@ -1,0 +1,81 @@
+/**
+ * Structured logging utilities for Yellowwood.
+ * Uses console.log/warn/error with consistent formatting.
+ */
+
+import { getErrorDetails } from './errorTypes.js';
+
+export type LogLevel = 'debug' | 'info' | 'warn' | 'error';
+
+interface LogContext {
+  [key: string]: unknown;
+}
+
+// Sensitive keys that should be redacted from logs
+const SENSITIVE_KEYS = new Set(['token', 'password', 'apiKey', 'secret', 'accessToken', 'refreshToken']);
+
+/**
+ * Safely stringify values, handling circular references and sensitive data
+ */
+function safeStringify(value: unknown): string {
+  const seen = new WeakSet<object>();
+  try {
+    return JSON.stringify(
+      value,
+      (key, val) => {
+        // Redact sensitive keys
+        if (SENSITIVE_KEYS.has(key)) return '[redacted]';
+
+        // Handle BigInt
+        if (typeof val === 'bigint') return val.toString();
+
+        // Handle circular references
+        if (val && typeof val === 'object') {
+          if (seen.has(val as object)) return '[Circular]';
+          seen.add(val as object);
+        }
+
+        return val;
+      },
+      2
+    );
+  } catch (error) {
+    // Fallback if JSON.stringify fails
+    return `[Unable to stringify: ${String(error)}]`;
+  }
+}
+
+/**
+ * Log a debug message (development only, filtered in production)
+ */
+export function logDebug(message: string, context?: LogContext): void {
+  if (process.env.NODE_ENV === 'development' || process.env.YELLOWWOOD_DEBUG) {
+    console.log(`[DEBUG] ${message}`, context ? safeStringify(context) : '');
+  }
+}
+
+/**
+ * Log an info message
+ */
+export function logInfo(message: string, context?: LogContext): void {
+  console.log(`[INFO] ${message}`, context ? safeStringify(context) : '');
+}
+
+/**
+ * Log a warning message
+ */
+export function logWarn(message: string, context?: LogContext): void {
+  console.warn(`[WARN] ${message}`, context ? safeStringify(context) : '');
+}
+
+/**
+ * Log an error message
+ */
+export function logError(message: string, error?: unknown, context?: LogContext): void {
+  const errorDetails = error ? getErrorDetails(error) : undefined;
+  console.error(
+    `[ERROR] ${message}`,
+    errorDetails ? safeStringify(errorDetails) : '',
+    context ? safeStringify(context) : ''
+  );
+}

